Add documentation for disable-force-detach-on-timeout (#45119)
kubernetes/website: content/en/docs/concepts/architecture/nodes.md

  recovered since the user was the one who originally added the taint.
{{< /note >}}

### Forced storage detach on timeout {#storage-force-detach-on-timeout}

> In any situation where a pod deletion has not succeeded for 6 minutes, kubernetes will

Reviewer: The "6 minutes" is a constant?

Author: Yeah, it's a hard-coded constant.

> force detach unmounting volumes if the node is unhealthy at that instant. Any

Reviewer: What is a "unmounting volume"? A volume being unmounted?

workload still running on the node that uses a force-detached volume will cause a
violation of the
[CSI specification](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerunpublishvolume),
which states that ControllerUnpublishVolume "MUST be called after all
NodeUnstageVolume and NodeUnpublishVolume on the volume are called and succeed".
In such circumstances, volumes on the node in question might encounter data corruption.

With the advent of the "Non-graceful node shutdown" feature, users might opt to disable
this behavior.

Force storage detach on timeout can be disabled by setting the `disable-force-detach-on-timeout`

> flag in `kube controller manager`; this means that a volume that is hosted on a node that

Reviewer: ```suggestion
flag in `kube-controller-manager`. Disabling the forced detach on timeout feature
means that a volume that is hosted on a unhealthy node
```

is unhealthy for more than 6 minutes will not have its associated

> [Volume Attachment](/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/)

Reviewer: ```suggestion
[VolumeAttachment](/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/)
```

  recovered since the user was the one who originally added the taint.
{{< /note >}}

> ### Forced storage detach on timeout {#storage-force-detach-on-timeout}

Reviewer 2: This is a lot of detail for someone who may be learning, for the very first time, what a Kubernetes cluster is.

We'd do well to find a better home for this detail.

Reviewer 2: How about https://kubernetes.io/docs/concepts/cluster-administration/?

Author: I don't know...

I feel like this belongs in the same section as [Non-Graceful node shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown).

After all, we're saying "This option is risky and can lead to data corruption; how about using the option above?"

Reviewer 2: > I feel like this belongs in the same section as [Non-Graceful node shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown).

I agree. We should move _both_ of those details to another page, ideally.

Author: I believe that that page was probably put there because graceful node shutdown was a thing:

https://github.com/kubernetes/website/blob/snapshot-initial-v1.20/content/en/docs/concepts/architecture/nodes.md#graceful-node-shutdown-graceful-node-shutdown

If we think this is the right thing to do, I'd be happy to move both Graceful & Non-Graceful, just Non-Graceful, or just the new section.

What makes the most sense at a high level?

Reviewer 3: If this merges, I've created an issue to move "Graceful Node Shutdown" and "Ungraceful Node Shutdown" sections along with "Forced storage detach on timeout" (PR pending) section
Issue https://github.com/kubernetes/website/issues/45690

In any situation where a pod deletion has not succeeded for 6 minutes, kubernetes will
force detach volumes being unmounted if the node is unhealthy at that instant. Any
workload still running on the node that uses a force-detached volume will cause a
violation of the
[CSI specification](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerunpublishvolume),

> which states that ControllerUnpublishVolume "MUST be called after all

Reviewer 2: ```suggestion
which states that `ControllerUnpublishVolume` "**must** be called after all
```
?

> NodeUnstageVolume and NodeUnpublishVolume on the volume are called and succeed".

Reviewer 2: ```suggestion
`NodeUnstageVolume` and `NodeUnpublishVolume` on the volume are called and succeed".
```
?

In such circumstances, volumes on the node in question might encounter data corruption.

> With the advent of the "Non-graceful node shutdown" feature, users might opt to disable
> this behavior.

Reviewer 2: Try:
```suggestion
The forced storage detach behaviour is optional.
```

Author: I've reworded this a bit.

The point is to suggest that folks use "Non-graceful node shutdown" instead of this.

Author: Please feel free to take another look.

Force storage detach on timeout can be disabled by setting the `disable-force-detach-on-timeout`
config field in `kube-controller-manager`. Disabling the force detach on timeout feature means
that a volume that is hosted on a node that is unhealthy for more than 6 minutes will not have
its associated
[VolumeAttachment](/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/)
deleted.

After this setting has been applied, unhealthy pods still attached to a volumes must be recovered

> via the [Non-Graceful Node Shutdown](#non-graceful-node-shutdown) procedure mentioned above.

Reviewer 4: Can you add a note that the user needs to use caution when using the "non graceful node shutdown" feature?  See the "Note" section of https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown.  Otherwise, they could still encounter data corruption.

Author: Thanks; I'll try to add something to that effect.

[CSI specification](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerunpublishvolume),
which states that ControllerUnpublishVolume "MUST be called after all
NodeUnstageVolume and NodeUnpublishVolume on the volume are called and succeed".
In such circumstances, volumes on the node in question might encounter data corruption.

With the advent of the "Non-graceful node shutdown" feature, users might opt to disable
this behavior.

Force storage detach on timeout can be disabled by setting the `disable-force-detach-on-timeout`
flag in `kube controller manager`; this means that a volume that is hosted on a node that
is unhealthy for more than 6 minutes will not have its associated
[Volume Attachment](/docs/reference/kubernetes-api/config-and-storage-resources/volume-attachment-v1/)
deleted.

However, manual mitigations might be required to repair the pod in question:
* The user might opt to manually taint the underlying node via the
[Non-Graceful Node Shutdown](#non-graceful-node-shutdown) procedure mentioned above if the node is
truly detached.
* If using the Non-Graceful Node Shutdown procedure is not an option, the user the user should

> opt to reboot the node in question.

Reviewer 5: I don't like presenting these as 2 alternatives, because the recommended mitigation is probably some combination of both. A reboot that brings the node back quickly is one option, but the more realistic option is to both shut down the node and apply the taint, then try to bring the node back.

I am very sensitive to any language which suggests that it's okay to apply the taint without actually shutting the node down, because that can still lead the problems. That misunderstanding can cause data corruption by itself, so we should reinforce the idea that the taint can only be applied after shutting down the node.

I propose this language:

There are two recommended options for cleaning up an unhealthy pod which is attached to a volume:
* Ensure the node is shut down, then taint the node according to the
[Non-Graceful Node Shutdown](#non-graceful-node-shutdown) procedure mentioned above.
* Reboot the node and wait for it to rejoin the cluster as a healthy node.

Author: > There are two recommended options for cleaning up an unhealthy pod which is attached to a volume:

If we're not giving the user "alternatives", then I don't think we should be giving them "options" either. Moreover, I think the steps are probably adequately captured in the non-graceful node shutdown above.

I think I'll just add a pointer to that section and indicate that that's the only way to fix the pod.
